cmd/kcp-front-proxy/authentication/groups.go

```diff
@@ -0,0 +1,138 @@
+/*
+Copyright 2022 The KCP Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package authentication
+
+import (
+	"fmt"
+	"net/http"
+	"strings"
+
+	"github.com/spf13/pflag"
+
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apiserver/pkg/authentication/authenticator"
+	"k8s.io/apiserver/pkg/authentication/request/x509"
+	"k8s.io/apiserver/pkg/authentication/user"
+	genericapiserver "k8s.io/apiserver/pkg/server"
+	apiserveroptions "k8s.io/apiserver/pkg/server/options"
+)
+
+// GroupFilter is a filter that filters out group that are not in the allowed groups,
+// and groups that are in the disallowed groups.
+type GroupFilter struct {
+	Authenticator authenticator.Request
+
+	AllowedGroups    sets.String
+	DisallowedGroups sets.String
+
+	AllowedGroupPrefixes    []string
+	DisallowedGroupPrefixes []string
+}
+
+var _ authenticator.Request = &GroupFilter{}
+
+func (a *GroupFilter) AuthenticateRequest(req *http.Request) (*authenticator.Response, bool, error) {
+	resp, ok, err := a.Authenticator.AuthenticateRequest(req)
+	if resp == nil || resp.User == nil {
+		return resp, ok, err
+	}
+
+	info := user.DefaultInfo{
+		Name:  resp.User.GetName(),
+		UID:   resp.User.GetUID(),
+		Extra: resp.User.GetExtra(),
+	}
+	resp.User = &info
+
+	groupsToPassOn := sets.NewString(resp.User.GetGroups()...)
+	if len(a.AllowedGroups) > 0 || len(a.AllowedGroupPrefixes) > 0 {
+	nextGroup:
+		for _, g := range groupsToPassOn.UnsortedList() {
+			if !a.AllowedGroups.Has(g) {
+				for _, prefix := range a.AllowedGroupPrefixes {
+					if strings.HasPrefix(g, prefix) {
+						goto nextGroup
+					}
+				}
+				groupsToPassOn.Delete(g)
```

Comment:
ok, so `allowedGroups` has precedence over `disallowGroups`. we should document this, both in gdoc an CLI help.

Reply:
good question. Clarified: `Disallowed groups trump over allowed groups.`